### PR TITLE
docs: rework setup instructions

### PR DESCRIPTION
**Pull request by jhchabran**
While we were pairing with @mrnugget on `sg setup`, it became clear that the current documentation is quite confusing for newcomers. 

This set of changes reorganise the getting started guide for local development. 

Goals:

- simplify those pages which have became hairy over time
- provide a place to put alternative setup instructions

Non-goals:

- reorganising the development section

Notable changes:

- Local Env is now a root section
  - mixing development how-tos with local env how-tos is confusing and contributed to a sense of mess for newcomers
  - a clear separation between the two will make upcoming contributions much easier.
- A page listing alternative options to install databases and languages have been added
- A ton of links have been updated and many links in the handbook will need to be updated.

How to review this:

- check out the branch locally
- `sg run docsite` 
- browse the new section

<details>
  <summary>Screenshot</summary>
  
CleanShot 2021-11-20 at 00 02 36@2x

</details>

### CODE REVIEW

File: doc/local-dev-env/index.md [hunk 1]
```diff
@@ -0,0 +1,80 @@
+# Local Development Environment
```
Comment (bobheadxi) [inline doc/local-dev-env/index.md:1]:
```suggestion
# Local development
```

Might I suggest we use something like `/doc/dev/env`?

- Most of docs.sourcegraph.com is product documentation, and having everything under `doc/dev` for dev docs is a nice delineation for customers to say "hey, you don't need to look at this"
- Given the above, since `dev` is in the path, we no longer need `dev` in the folder name
- We don't have a particular use case for remote development, so we can probably drop `local`

=> `/doc/dev/env`

I think shorter path names are a bit more ergonomic to share and reason with

File: doc/local-dev-env/alternative-set-up-for-languages.md [hunk 2]
```diff
@@ -0,0 +1,61 @@
+# Alternative set up for languages
```
Comment (bobheadxi) [inline doc/local-dev-env/alternative-set-up-for-languages.md:1]: I think we should just merge the `Alternative set up...` into a single page for alternative setups

File: doc/local-dev-env/how-to/index.md [hunk 3]
```diff
@@ -0,0 +1,20 @@
+# How to guides
+
+- [How to debug live code](debug_live_code.md)
+- [Set up local development with Zoekt and Sourcegraph](zoekt_local_dev.md)
+- [Ignoring editor config files in Git](ignoring_editor_config_files.md)
+- [Use `golangci-lint`](use-golangci-lint.md)
+- [Set up local Sourcegraph monitoring development](monitoring_local_dev.md)
+
+## Offline development
```
Comment (bobheadxi) [inline doc/local-dev-env/how-to/index.md:9]: Can we move this alongside https://docs.sourcegraph.com/dev/background-information/sg#attach-a-debugger?

File: doc/local-dev-env/index.md (hunk 1 above)
Comment (jhchabran) [reply doc/local-dev-env/index.md:1]: > "hey, you don't need to look at this"

I've been asking myself this question for a while when I was making those changes as well. The main issue I couldn't find a solution for is how can we present the local environment stuff as an easily found bucket of resources for both employees and external contributors. 

- We can't really do anything in the sidebar to address that if we nest it under the development section.
- Potentially just me stretching the reasoning too much, but it also gives some insights on how committed we are to open source contributions.
- I don't think there is any kind of confusion for a customer regarding, especially since it's below the development section. Still your point stands, it's a bit more noise for them. 
  - So far, I feel like it's a fair price to pay for the added clarity. 

> => /doc/dev/env

That's a very good remark, if the menu highlighting is not broken, we can definitely do this. And if not, we could still rename the folder into something short such as `local`.

Comment (bobheadxi) [reply doc/local-dev-env/index.md:1]: > The main issue I couldn't find a solution for is how can we present the local environment stuff as an easily found bucket of resources for both employees and external contributors.

I think featuring it prominently in `doc/dev/index.md`, and making `doc/dev/index.md` the "homepage for contributing to Sourcegraph" is the way to go. I'm not a fan of the current "giant list of table of contents" format - there should be a bit more design and flow to our docs, especially landing pages like `doc/dev/index.md`. We should also link to `doc/dev/index.md` from places like our `README.md` and `CONTRIBUTING.md` as the first step.

Basically, landing in `doc/dev/index.md` from "Development" in the sidebar should point to everything that you need to get started, including links to resources to set up your development environment (I really don't think we need to specify _local_) and get a preliminary overview of how Sourcegraph works.

> it also gives some insights on how committed we are to open source contributions.

I think making `doc/dev/index.md` super friendly to OSS is the way to go. Anything that's _not_ for OSS should be in the handbook, and noted as such in any `doc/` content

Comment (jhchabran) [reply doc/local-dev-env/index.md:1]: > I'm not a fan of the current "giant list of table of contents" format 

Me neither, it could really be improved. But I don't want to touch this within this PR, it's going to become too large otherwise. 

> Basically, landing in doc/dev/index.md from "Development" in the sidebar should point to everything that you need to get started, including links to resources to set up your development environment

I don't understand the point you're making here. Isn't this what we had before? 

The Development section is conflating the local environment setup with general development. So you end up with two groups of how-tos, and it's really hard to scan the page because you can find references to a particular component in both groups. 

Also, it's not just software engineers that want to run the local env, product designers, PMs and others need to, so this page has to be taking their needs in account.

> Unfortunately I feel this is a loss in clarity - where before we can just go to doc/dev for everything related to contributing to Sourcegraph, now that is spread out in a manner that blurs the dilineation between dev docs and product docs

I understand your point. I don't have that strong opinions on whether or not it should be a top level section, the part I'm really after is having a clear overview of what you can do to setup the local environment and avoid at all cost to mix this with the rest of the development pages. 

➡️ We could move it to `doc/dev/local` as you suggested and add a very visible "Local Env Dev" thing at the top of the development page.

File: doc/local-dev-env/how-to/index.md (hunk 3 above)
Comment (jhchabran) [reply doc/local-dev-env/how-to/index.md:9]: Good idea 👍

File: doc/local-dev-env/index.md (hunk 1 above)
Comment (bobheadxi) [reply doc/local-dev-env/index.md:1]: > We could move it to doc/dev/local(or local_env) as you suggested and add a very visible "Local Env Dev" thing at the top of the development page.

+1 to this! 😁

Comment (mrnugget) [reply doc/local-dev-env/index.md:1]: > I think featuring it prominently in doc/dev/index.md, and making doc/dev/index.md the "homepage for contributing to Sourcegraph" is the way to go.

+1000 on this. The more I work on `sg setup` the more I think "we need better docs" that people can actually use without knowing which exact document they need.

Comment (mrnugget) [reply doc/local-dev-env/index.md:1]: > > => /doc/dev/env
>
> That's a very good remark, if the menu highlighting is not broken, we can definitely do this. And if not, we could still rename the folder into something short such as local.
> @mrnugget any thoughts on this on your side now you've slept on it?

I thought about this too and I'm wondering whether `env` is the best word to use here. I'm thinking of `setup` for example. That's what I would look for when I go to `doc/dev` and have no local env running: how do I setup, what is the setup, etc.

File: doc/local-dev-env/alternative-set-up-for-databases.md [hunk 4]
```diff
@@ -0,0 +1,34 @@
+# Alternative set up for databases
+
+The sourcegraph codebase requires a few databases to be installed in order to run locally. We recommend and support a default approach that is suited for everyone, regardless of their familiarity with the topic.
```
Comment (mrnugget) [inline doc/local-dev-env/alternative-set-up-for-databases.md:3]:
```suggestion
Sourcegraph requires databases to be accessible to run locally. We recommend and support a default approach that is suited for everyone, regardless of their familiarity with the topic. 
```

File: doc/local-dev-env/alternative-set-up-for-databases.md [hunk 5]
```diff
@@ -0,0 +1,34 @@
+# Alternative set up for databases
+
+The sourcegraph codebase requires a few databases to be installed in order to run locally. We recommend and support a default approach that is suited for everyone, regardless of their familiarity with the topic. 
+
+It's totally possible to use alternative ways to install databases if you need to: `sg` checks will still perform correctly, but you're on your own for the installation process: see the [alternative instructions section](#alternative-instructions).
```
Comment (mrnugget) [inline doc/local-dev-env/alternative-set-up-for-databases.md:5]:
```suggestion
But it's still possible to use alternative ways to install databases if you want to and know what you're doing: `sg checks` will still perform correctly, but you're on your own for the installation process: see the [alternative instructions section](#alternative-instructions).
```

File: doc/local-dev-env/alternative-set-up-for-databases.md [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+
+## MacOs
```
Comment (mrnugget) [inline doc/local-dev-env/alternative-set-up-for-databases.md:7]:
```suggestion
## macOS
```

File: doc/local-dev-env/alternative-set-up-for-databases.md [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+
+- Install _Postgres.app_:
+  1. Open your browser and navigate to this page: [https://postgresapp.com](https://postgresapp.com)
+  1. Follow the installation instructions.
+- Install _Redis.app_:
+  1. Open your browser and navigate to this page: [https://jpadilla.github.io/redisapp/](https://jpadilla.github.io/redisapp/)
+  1. Click on the Download button and install the app in the archive.
+  1. Open a terminal and type: `sudo mkdir -p /etc/paths.d && echo /Applications/Redis.app/Contents/Resources/Vendor/redis/bin | sudo tee /etc/paths.d/redisapp`
+
+### Using Homebrew
```
Comment (mrnugget) [inline doc/local-dev-env/alternative-set-up-for-databases.md:26]:
```suggestion
### Using Homebrew (recommended!)
```

File: doc/local-dev-env/alternative-set-up-for-databases.md [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+It's totally possible to use alternative ways to install databases if you need to: `sg` checks will still perform correctly, but you're on your own for the installation process: see the [alternative instructions section](#alternative-instructions).
+
+## MacOs
+
+### Using packaged applications
+
+Requirements: 
+
+- The `github.com/sourcegraph/sourcegraph` repository cloned in a folder of your choice.
+- [`Homebrew`](https://brew.shell) is installed.
```
Comment (mrnugget) [inline doc/local-dev-env/alternative-set-up-for-databases.md:14]: I don't think this needs to be here? It's general and applies to a lot of things, but it _doesn't_ apply to `Postgres.app` and `Redis.app`.

File: doc/local-dev-env/alternative-set-up-for-databases.md (hunk 7 above)
Comment (mrnugget) [reply doc/local-dev-env/alternative-set-up-for-databases.md:26]: I think we should make this the recommended way since it's the easiest to automatically run with `sg setup` and also the easiest from a "get `psql` into `$PATH`" perspective

File: doc/local-dev-env/alternative-set-up-for-languages.md (hunk 2 above)
Comment (mrnugget) [reply doc/local-dev-env/alternative-set-up-for-languages.md:1]: Agree. Especially since the first two paragraphs are nearly duplicates.

Stepping back a bit: I'd love to have a single, well-structured, well-formatted page that (a) lists all dependencies and (b) shows how to install them and (c) which alternative install methods are available. I think it's possible to create such a page without it becoming unwieldy and it would be an amazing document to have because we could link to it.

File: doc/local-dev-env/alternative-set-up-for-languages.md [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+
+The sourcegraph codebase requires a few languages to be installed in orde to run locally. We recommend and support a default approach that is suited for everyone, regardless of their familiarity with the topic. 
+
+It's totally possible to use alternative ways to install languages if you need to: `sg` checks will still perform correctly, but you're on your own for the installation process: see the [alternative instructions section](#alternative-instructions).
+
+## MacOs
+
+### Using `asdf` for everything
```
Comment (mrnugget) [inline doc/local-dev-env/alternative-set-up-for-languages.md:9]: If this is our recommended way, then I'm a bit confused that it's under "alternatives", see comment above.

**Review by mrnugget — COMMENTED** [review]
See comments. I'm not sure wehther `local-dev-env` is the best possible name. I also think that maybe we should come up with an organization scheme/folder structure and naming _before_ we move things around?

Because if we have that it's much easier to review/approve these PRs.

File: doc/local-dev-env/index.md (hunk 1 above)
Comment (bobheadxi) [reply doc/local-dev-env/index.md:1]: `doc/dev/setup` sounds like a good home for this! It aligns nicely with `sg setup` as well 😁

Comment (jhchabran) [reply doc/local-dev-env/index.md:1]: Alright, let's do this! I'll move this around tomorrow plus a more visible highlight in the dev docs we'll get this merged 😊

File: doc/local-dev-env/alternative-set-up-for-languages.md [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+
+1. Open a terminal and type: 
+1. `brew install asdf`
+1. `echo ' . /opt/homebrew/opt/asdf/libexec/asdf.sh' >> ~/.zshrc:`
+1. `zsh`
+1. `asdf version`
+  - this should print something similar to `v0.8.1` (the numbers are not important) 
+  - if you get `zsh: command not found: asdf` then something did not work.
+  <!--- TODO replace this with `sg setup2 checks -->
+1. `asdf plugin add golang`
```
Comment (ggilmore) [inline doc/local-dev-env/alternative-set-up-for-languages.md:26]: This part could be scripted (use awk to parse `.tool-versions` instead of manually adding each plugin)

File: doc/local-dev-env/alternative-set-up-for-languages.md (hunk 9 above)
Comment (jhchabran) [reply doc/local-dev-env/alternative-set-up-for-languages.md:9]: I kept it this way to not end up having inconsistencies with the current text in `sg setup`. We can address this in a different PR.

**Comment by jhchabran** [comment]
Updated as per the comments here and in the working doc at https://docs.google.com/document/d/1nPGy6FOc8oFXhilGA7CuUu2MHlDHB4DuiXNs-mOADN8/edit# 

Right now, I'm a bit wondering on the two refs to Quickstart 👇. I think ideally, it should be "Contribute" in there, but our current page for that is quite incomplete on the backend. 

CleanShot 2021-11-25 at 08 21 09@2x

File: doc/dev/index.md [hunk 11]
```diff
@@ -1,17 +1,70 @@
...
+  </a>
+
+  <a href="setup/troubleshooting" class="btn" alt="Troubleshooting">
+   <span>Troubleshooting</span>
+   </br>
+  List of the <b>most common</b> local environment problems and their solutions.
+  </a>
+</div>
+
+<br/>
```
Comment (mrnugget) [inline doc/dev/index.md:67]: I didn't want to push to your branch, but here's what I did locally:

screenshot_2021-11-25_09 01 26@2x


```suggestion
<p class="subtitle">Documentation for <b>developers contributing to the Sourcegraph code base</b></p>

<div class="cta-group">
<a class="btn btn-primary" href="setup/quickstart">★ Quickstart: develop Sourcegraph on your machine</a>
<a class="btn" href="https://github.com/sourcegraph/sourcegraph">GitHub repository</a>
<a class="btn" href="https://github.com/sourcegraph/sourcegraph/issues">Issue Tracker</a>
</div>


## [Setup](setup/index.md)

<p class="subtitle">Learn how to develop Sourcegraph on your machine.</p>

<div class="getting-started">
  <a href="setup/quickstart" class="btn" alt="Run through the Quickstart guide">
   <span>★ Quickstart</span>
   </br>
   Run through the <b>step by step guide</b> and get your local environment ready.
  </a>

  <a href="setup/how-to" class="btn" alt="How-to guides">
   <span>How-to guides</span>
   </br>
  <b>Context specific</b> guides: debugging live code, Apple M1 workarounds, ...
  </a>

  <a href="setup/troubleshooting" class="btn" alt="Troubleshooting">
   <span>Troubleshooting</span>
   </br>
  Help for the <b>most common</b> problems.
  </a>
</div>
```

**Review by mrnugget — APPROVED** [review]
Thanks for sticking with this! Let's get this in and then we can iterate and move details around etc. if necessary

File: doc/dev/index.md (hunk 11 above)
Comment (bobheadxi) [reply doc/dev/index.md:67]: This looks great 😍

Comment (mrnugget) [reply doc/dev/index.md:67]: We can and should then do the same cleanup for the other sections too. We have great docs but they're just really hard to find.
